docs/execution_providers/DirectML-ExecutionProvider.md

```diff
@@ -0,0 +1,93 @@
+
+## Table of contents
+
+- [DirectML Execution Provider](#directml-execution-provider)
+  - [Table of contents](#table-of-contents)
+  - [Minimum requirements](#minimum-requirements)
+  - [Building from source](#building-from-source)
+  - [Using the DirectML execution provider](#using-the-directml-execution-provider)
+    - [`OrtSessionOptionsAppendExecutionProvider_DML` function](#ortsessionoptionsappendexecutionproviderdml-function)
+    - [`OrtSessionOptionsAppendExecutionProviderEx_DML` function](#ortsessionoptionsappendexecutionproviderexdml-function)
+    - [ONNX opset support](#onnx-opset-support)
+  - [Samples](#samples)
+  - [See also](#see-also)
+
+## Minimum requirements
+
+The DirectML execution provider requires any DirectX 12 capable device. Almost all commercially-available graphics cards released in the last several years support DirectX 12. Examples of compatible hardware include:
+
+* NVIDIA Kepler (GTX 600 series) and above
+* AMD GCN 1st Gen (Radeon HD 7000 series) and above
+* Intel Haswell (4th-gen core) HD Integrated Graphics and above
+
+DirectML is compatible with Windows 10, version 1709 (10.0.16299; RS3, "Fall Creators Update") and newer.
+
+
+
+## Building from source
+
+For general information about building onnxruntime, see [BUILD.md](../../BUILD.md).
+
+Requirements for building the DirectML execution provider:
+1. Visual Studio 2017 toolchain (see [cmake configuration instructions](../../BUILD.md))
+2. [The Windows 10 SDK (10.0.18362.0) for Windows 10, version 1903](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk) (or newer)
+
+To build onnxruntime with the DML EP included, supply the `--use_dml` parameter to `build.bat`. e.g.
+
+    build.bat --config RelWithDebInfo --build_shared_lib --parallel --use_dml
+
+The DirectML execution provider supports building for both x64 (default) and x86 architectures.
+
+
+
+## Using the DirectML execution provider
+
+When using the [C API](../C_API.md) with a DML-enabled build of onnxruntime (see [Building from source](#building-from-source)), the DirectML execution provider can be enabled using one of the two factory functions included in `include/onnxruntime/core/providers/dml/dml_provider_factory.h`.
+
+### `OrtSessionOptionsAppendExecutionProvider_DML` function
+
+ Creates a DirectML Execution Provider which executes on the hardware adapter with the given device_id, also known as the adapter index. The device ID corresponds to the enumeration order of hardware adapters as given by [IDXGIFactory::EnumAdapters](https://docs.microsoft.com/windows/win32/api/dxgi/nf-dxgi-idxgifactory-enumadapters). A device_id of 0 always corresponds to the default adapter, which is typically the primary display GPU installed on the system. A negative `device_id` is invalid.
+
+     OrtStatus* OrtSessionOptionsAppendExecutionProvider_DML(
+        _In_ OrtSessionOptions* options,
+        int device_id
```

Comment:
Do we want to stick with enumertion ID?   or do we want to have LUID's as currencies?    if we stick with device_id as an enumeration id , will it work with DXCORE?

Reply:
also, will passing 0 give us the WARP adapter?  or will we always get a hardware adapter and this api will fail if there are no GPU's in the system?

---
In reply to: [333245308](https://github.com/microsoft/onnxruntime/pull/2057#discussion_r333245308) [](ancestors = 333245308)

Reply:
`device_id` was chosen to be consistent with the other EPs. e.g. the CUDA EP also has the notion of an integer device ID. For the DML EP, the `device_id` corresponds to the DXGI adapter enumeration index. A `device_id` of zero always corresponds to the default adapter, which is typically the primary display GPU installed on the system. If the system has no GPU (e.g. if it's a VM) then the primary display "GPU" is WARP.